include/rdma/fi_ext.h

```diff
@@ -54,11 +54,11 @@ extern "C" {
  * }
  */
 
-#define FI_PROV_SPECIFIC_EFA   (FI_PROV_SPECIFIC | 0xefa << 20)
+#define FI_PROV_SPECIFIC_EFA   (FI_PROV_SPECIFIC | 0xefaU << 20)
 
 /* EFA options */
 enum {
-       FI_OPT_EFA_RNR_RETRY = (1U | FI_PROV_SPECIFIC_EFA),
+       FI_OPT_EFA_RNR_RETRY = (int)(1U | FI_PROV_SPECIFIC_EFA),
```

Comment:
ewww... I guess the integer cast has to happen somewhere (since setopt() takes an int), but that's an unfortunate set of choices we have.

Reply:
Can we solve part of this by removing the Unsigned integers in this flow?  Even FI_PROV_SPECIFIC tacks on the U for some reason.  Does redefining it (but keeping the same resulting value) help?

Reply:
I'm really bad at math.  `0xefa << 20` results in the 31st bit being `1`.  `0xefa` is 12 bits long with the high bit being 1 (because `0xe`), and then we bit shift it 20 bits, so  we end up with the 31st bit set.  C99 says that will be promoted to a long int.  `0xefaU` worked around that, because there are 32 data bits for unsigned, so it doesn't promote to long.

Likewise, `FI_PROV_SPECIFIC` is a problem, because `1U << 31` has type unsigned int and `1 << 31` is a long int.  So after type promotion of all the constants, you end up with either a long int or unsigned int expression, both of which are not integer constants.

There are two options that would work without changing the constant value:

```
#define FI_PROV_SPECIFIC       ((int)(1U << 31))
#define FI_PROV_SPECIFIC_EFA   (FI_PROV_SPECIFIC | (0x6fa << 20))
enum {
       FI_OPT_EFA_RNR_RETRY = FI_PROV_SPECIFIC_EFA,
};
```

or 

```
#define FI_PROV_SPECIFIC       (1U << 31)
#define FI_PROV_SPECIFIC_EFA   ((int)FI_PROV_SPECIFIC | (0x6fa << 20))
enum {
       FI_OPT_EFA_RNR_RETRY = FI_PROV_SPECIFIC_EFA,
};
```

Note that I changed the provider constant from `0xefa` to `0x6fa` to avoid the leading bit problem in the first paragraph.  I know Sean had a bit allocation, but I can't find it right now.  We need to adjust the constant to fit Sean's allocation in the real patch.

I don't have strong preferences between the two possible fixes.  The first has a broader blast radius, although maybe it doesn't matter.

Reply:
From what I understood of C standard, for `enumerator`, in the form of `enumeration-constant  = constant-expression`, `constant-expression` has to be integer constant expression whose value is representable as a value of type `int`.

To have a valid integer constant expression, we need to avoid overflowing the 1 to the sign bit. For example, we use `1U << 31`, `0xefaU << 20`. We can also use suffix `L`, or some other ways. But to have a value that is representable as type `int`, we still need cast the value of the integer constant expression to `int`.   

I could not find a way to remove the cast by redefining the macro but keeping the same resulting value. 

Is it acceptable if we define `FI_PROV_SPECIFIC_XXX` without touching the sign bit, and without using `FI_PROV_SPECIFIC`, like just
```
FI_PROV_SPECIFIC_EFA   (0xefa << 19)
```

Reply:
Maybe simpler is better here then:
```
/* negative options are provider specific */

enum {
    FI_OPT_EFA = -(0xefa << 16),
    FI_OPT_EFA_RNR_RETRY,
    FI_OPT_EFA_CALL_MOM,
    ...
};
```

Reply:
>So after type promotion of all the constants, you end up with either a long int or unsigned int expression, both of which are not integer constants.

@bwbarrett I'm not sure of this, I could be wrong here though. I read the definition of [Integer constant expression](https://en.cppreference.com/w/c/language/constant_expression) and [integer constants](https://en.cppreference.com/w/c/language/integer_constant). 

My understanding is that the `integer constant expression` consists of `integer constants`, and `integer constants` are allowed to have integer-suffix. This means, with integer-suffix, `U`, `L`, or `LL`, we will NOT get warning `enumerator value for xxx is not an integer constant expression [-Wpedantic]`. I verified this. So I'm not sure if the promotion (`1 << 31` promoted to long int) really happens, or if the promotion has the same effect as adding integer-suffix `L`. 

But we still get the warning `ISO C restricts enumerator values to range of ‘int’ [-Wpedantic]`, because neither long int nor unsigned int is in the range of `int`, so we need to cast to `int`.

Personally, I'm inclined to @shefty's solution, if we do not need to rely on the macro `FI_PROV_SPECIFIC` any more.

Reply:
@shefty's suggestion works for me.